driver-core/src/main/com/mongodb/operation/CommandOperationHelper.java

```diff
@@ -391,6 +421,158 @@ public void onResult(final D result, final Throwable t) {
                 });
     }
 
+    /* Retryable write helpers */
+    static <T, R> R executeRetryableCommand(final WriteBinding binding, final String database,
+                                            final FieldNameValidator fieldNameValidator, final Decoder<T> commandResultDecoder,
+                                            final CommandCreator commandCreator,
+                                            final CommandTransformer<T, R> transformer) {
+        return withReleasableConnection(binding, new CallableWithConnectionAndSource<R>() {
+            @Override
+            public R call(final ConnectionSource source, final Connection connection) {
+                BsonDocument command = null;
+                MongoException exception = null;
+                try {
+                    command = commandCreator.create(source.getServerDescription(), connection.getDescription());
+                    return transformer.apply(connection.command(database, command, fieldNameValidator, ReadPreference.primary(),
+                            commandResultDecoder, binding.getSessionContext()), connection.getDescription().getServerAddress());
+                } catch (MongoException e) {
+                    exception = e;
+                    if (command == null || !command.containsKey("txnNumber") || !isRetryableException(exception)) {
+                        throw exception;
+                    }
+                } finally {
+                    connection.release();
+                }
+
+                final BsonDocument originalCommand = command;
+                final MongoException originalException = exception;
+                return withConnection(binding, new OperationHelper.CallableWithConnection<R>() {
```

Comment:
It looks like this is violating this part of the spec:

    /* If we cannot select a writable server, do not proceed with retrying and
     * throw the original error. The caller can then infer that an attempt was
     * made and failed. */
    try {
      server = selectServer("writable");
    } catch (Exception ignoredError) {
      throw originalError;
    }

i.e. it's not throwing the original error if server selection fails.

Reply:
Done.